geode-core/src/main/java/org/apache/geode/management/internal/FederatingManager.java

```diff
@@ -380,8 +384,10 @@ void addMemberArtifacts(InternalDistributedMember member) {
         return;
       }
 
-      try {
+      FederatingManagerCancelCriterion cancelCriterion =
```

Comment:
Creating a new CancelCriterion instance isn't really what is needed here. You need to `getCancelCriterion` from either the Cache or the DistributedSystem and pass that into the constructor of the classes that are using it. Then if the DS or Cache are closing, the CancelCriterion will cause the FederatingManager to cancel with the cause being the Cache or DS being closed.

If you look at #5728 (my PR with changes to FederatingManager), you'll see that the constructors are now passing in an instance of `CancelCriterion` which comes from `InternalDistributedSystem`:

```
FederatingManager(ManagementResourceRepo repo, InternalDistributedSystem system,
      SystemManagementService service, InternalCache cache, StatisticsFactory statisticsFactory,
      StatisticsClock statisticsClock, MBeanProxyFactory proxyFactory, MemberMessenger messenger,
      Supplier<ExecutorService> executorServiceSupplier) {
    this(repo, system, service, cache, statisticsFactory, statisticsClock, proxyFactory, messenger,
        system.getCancelCriterion(), executorServiceSupplier);
  }

  @VisibleForTesting
  FederatingManager(ManagementResourceRepo repo, InternalDistributedSystem system,
      SystemManagementService service, InternalCache cache, StatisticsFactory statisticsFactory,
      StatisticsClock statisticsClock, MBeanProxyFactory proxyFactory, MemberMessenger messenger,
      CancelCriterion cancelCriterion, ExecutorService executorService) {
    this(repo, system, service, cache, statisticsFactory, statisticsClock, proxyFactory, messenger,
        cancelCriterion, () -> executorService);
  }

  @VisibleForTesting
  FederatingManager(ManagementResourceRepo repo, InternalDistributedSystem system,
      SystemManagementService service, InternalCache cache, StatisticsFactory statisticsFactory,
      StatisticsClock statisticsClock, MBeanProxyFactory proxyFactory, MemberMessenger messenger,
      CancelCriterion cancelCriterion, Supplier<ExecutorService> executorServiceSupplier) {
    super(repo, system, cache, statisticsFactory, statisticsClock);
    this.service = service;
    this.proxyFactory = proxyFactory;
    this.messenger = messenger;
    this.executorServiceSupplier = executorServiceSupplier;
    lifecycleLock = new StoppableReentrantLock(cancelCriterion);
  }
```

Reply:
I'm really not familiar with the usage of `CancelCriterion` so perhaps my understanding of what's required here is a bit naive... My understanding was simply that we didn't want to indefinitely block any threads `await`ing in either of the two listeners in question here. That situation would arise if something bad happened in the `FederatingManager` before it could get to calling `readyForEvents`. Hence the use of the `StoppableCountdownLatch`.

Since what's calling into these listeners is a callback using an asynchronous thread, I don't understand why it is the responsibility of these listeners to go through all of this ceremony to tell some random component that the cache is closing. Surely that component should have its own means of determining that the cache is closing? Why is it not enough for the thread to be released and throw an exception?

Reply:
ok, maybe it makes most sense to wait wait until 5278 is pull merged, and then adapt off that.  Are you expecting to have time to get that through in the near future?

Reply:
I hope to have #5728 ready to merge today at the soonest or in a day or two.

Reply:
@jdeppe-pivotal hmm, well now you have me second-guessing the need for CancelCriterion. Maybe we can get another review before we make more changes on this PR.